src/cdomain/value/cdomains/int/bitfieldDomain.ml

```diff
@@ -0,0 +1,623 @@
+      let (z,o) = t in
+      Format.sprintf "{%s, (zs:%s, os:%s)}" (to_pretty_bits t) (Ints_t.to_string z) (Ints_t.to_string o) 
+
+  include Std (struct type nonrec t = t let name = name let top_of = top_of let bot_of = bot_of let show = show let equal = equal end)
+
+  let range ik bf = (BArith.min ik bf, BArith.max ik bf)
+
+  let maximal (z,o) =     
+    if (z < Ints_t.zero) <> (o < Ints_t.zero) then Some o
+    else None
+
+  let minimal (z,o) =     
+    if (z < Ints_t.zero) <> (o < Ints_t.zero) then Some (!:z)
+    else None
+
+  let wrap ik (z,o) = 
+    let (min_ik, max_ik) = Size.range ik in
+    if isSigned ik then
+      let newz = (z &: (Ints_t.of_bigint max_ik)) |: ((Ints_t.of_bigint min_ik) *: (BArith.nth_bit z (Size.bit ik - 1))) in
+      let newo = (o &: (Ints_t.of_bigint max_ik)) |: ((Ints_t.of_bigint min_ik) *: (BArith.nth_bit o (Size.bit ik - 1))) in
+      (newz,newo)
+    else
+      let newz = z |: !:(Ints_t.of_bigint max_ik) in
+      let newo = o &: (Ints_t.of_bigint max_ik) in
+      (newz,newo)
+
+  let norm ?(suppress_ovwarn=false) ik (z,o) = 
+    if BArith.is_invalid (z,o) then 
+      (bot (), {underflow=false; overflow=false})
+    else      
+      let (min_ik, max_ik) = Size.range ik in
+      let isPos = z < Ints_t.zero in 
+      let isNeg = o < Ints_t.zero in
+      let underflow = if isSigned ik then (((Ints_t.of_bigint min_ik) &: z) <> Ints_t.zero) && isNeg else isNeg in
+      let overflow = (((!:(Ints_t.of_bigint max_ik)) &: o) <> Ints_t.zero) && isPos in
+      let new_bitfield = wrap ik (z,o)
+      in
+      let overflow_info = if suppress_ovwarn then {underflow=false; overflow=false} else {underflow=underflow; overflow=overflow} in      
+      if not (underflow || overflow) then
+        ((z,o), overflow_info)
+      else if should_wrap ik then
+        (new_bitfield, overflow_info)
+      else if should_ignore_overflow ik then 
+        (* (M.warn ~category:M.Category.Integer.overflow "Bitfield: Value was outside of range, indicating overflow, but 'sem.int.signed_overflow' is 'assume_none' -> Returned Top"; *)
+        (* (bot (), overflow_info)) *)
+        (top_of ik, overflow_info)
+      else 
+        (top_of ik, overflow_info)
+
+  let cast_to ?(suppress_ovwarn=false) ?torg ?no_ov ik x = norm ~suppress_ovwarn:(suppress_ovwarn || x = top ()) ik x
+
+  let join ik b1 b2 = (norm ik @@ (BArith.join b1 b2) ) |> fst
+
+  let meet ik x y = (norm ik @@ (BArith.meet x y)) |> fst
+
+  let leq (x:t) (y:t) = (BArith.join x y) = y
+
+  let widen ik x y = (norm ik @@ BArith.widen x y) |> fst
+
+  let narrow ik x y = meet ik x y
+
+  let of_int ik (x: int_t) = (norm ik @@ BArith.of_int x) 
+
+  let to_int (z,o) = if is_bot (z,o) then None else
+    if BArith.is_const (z,o) then Some o
+    else None
+
+  let equal_to i bf = 
+    if BArith.of_int i = bf then `Eq
+    else if leq (BArith.of_int i) bf then `Top
+    else `Neq
+
+  (* Conversions *)
+
+  let of_interval ?(suppress_ovwarn=false) ik (x,y) =
```

Comment:
Does this function share code with the `to_bitfield` function in the interval domain? If so, could this be deduplicated?